File: app/src/main/java/org/codepond/commitbrowser/common/ui/BaseActivity.java
package org.codepond.commitbrowser.common.ui;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import com.android.databinding.library.baseAdapters.BR;

import javax.inject.Inject;

import dagger.android.AndroidInjection;
import rx.Observable;
import rx.Subscription;

public abstract class BaseActivity<T extends ViewModel, R extends ViewDataBinding> extends AppCompatActivity {
    private T viewModel;
    private Subscription subscription;
    private R binding;
    private NetworkErrorSnackBar networkErrorSnackBar;
    @Inject ViewModelFactory viewModelFactory;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState);
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(getViewModelClass());
        binding = DataBindingUtil.setContentView(this, getLayoutId());
        binding.setVariable(BR.viewModel, viewModel);
        networkErrorSnackBar = new NetworkErrorSnackBar(binding.getRoot());
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (subscription != null && !subscription.isUnsubscribed()) {
            subscription.unsubscribe();
        }
    }

    protected abstract Class<T> getViewModelClass();

    @LayoutRes protected abstract int getLayoutId();

    protected <V extends Observable<?>> void subscribe(V observable) {
        subscription = observable
                .retryWhen(networkErrorSnackBar::show)
                .subscribe();
    }

    public T getViewModel() {
        return viewModel;
    }

    public R getBinding() {
        return binding;
    }
}

File: app/src/main/java/org/codepond/commitbrowser/common/ui/BaseActivity.kt
package org.codepond.commitbrowser.common.ui

import android.os.Bundle
import androidx.annotation.LayoutRes
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.databinding.library.baseAdapters.BR
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProviders
import dagger.android.support.DaggerAppCompatActivity
import org.codepond.commitbrowser.di.ViewModelFactory
import javax.inject.Inject

abstract class BaseActivity<T : ViewModel, R : ViewDataBinding> : DaggerAppCompatActivity() {
    @Inject
    lateinit var viewModelFactory: ViewModelFactory

    protected val viewModel: T by lazy(LazyThreadSafetyMode.NONE) {
        ViewModelProviders.of(this, viewModelFactory).get(viewModelClass)
    }

    protected val binding: R by lazy(LazyThreadSafetyMode.NONE) {
        DataBindingUtil.setContentView<R>(this, layoutId)
    }

    protected abstract val viewModelClass: Class<T>

    @get:LayoutRes
    protected abstract val layoutId: Int

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding.lifecycleOwner = this
        binding.setVariable(BR.viewModel, viewModel)
    }
}
